pkg/controller/networkpolicy/status_controller.go

```diff
@@ -0,0 +1,343 @@
+	}
+	key := internalNetworkPolicyKeyFunc(oldANP)
+	c.queue.Add(key)
+}
+
+func (c *StatusController) UpdateStatus(status *controlplane.NetworkPolicyStatus) error {
+	key := status.Name
+	_, found, _ := c.internalNetworkPolicyStore.Get(key)
+	if !found {
+		klog.Infof("NetworkPolicy %s has been deleted, skip updating its status", key)
+		return nil
+	}
+	func() {
+		c.statusesLock.Lock()
+		defer c.statusesLock.Unlock()
+		statusPerNode, exists := c.statuses[key]
+		if !exists {
+			statusPerNode = map[string]*controlplane.NetworkPolicyNodeStatus{}
+			c.statuses[key] = statusPerNode
+		}
+		for i := range status.Nodes {
+			statusPerNode[status.Nodes[i].NodeName] = &status.Nodes[i]
+		}
+	}()
+	c.queue.Add(key)
+	return nil
+}
+
+func (c *StatusController) getNodeStatuses(key string) []*controlplane.NetworkPolicyNodeStatus {
+	c.statusesLock.RLock()
+	defer c.statusesLock.RUnlock()
+	statusPerNode, exists := c.statuses[key]
+	if !exists {
+		return nil
+	}
+	statuses := make([]*controlplane.NetworkPolicyNodeStatus, 0, len(c.statuses[key]))
+	for _, status := range statusPerNode {
+		statuses = append(statuses, status)
+	}
+	return statuses
+}
+
+func (c *StatusController) clearStatuses(key string) {
+	c.statusesLock.Lock()
+	defer c.statusesLock.Unlock()
+	delete(c.statuses, key)
+}
+
+func (c *StatusController) deleteNodeStatus(key string, nodeName string) {
+	c.statusesLock.Lock()
+	defer c.statusesLock.Unlock()
+	statusPerNode, exists := c.statuses[key]
+	if !exists {
+		return
+	}
+	delete(statusPerNode, nodeName)
+}
+
+// Run begins watching and syncing of a StatusController.
+func (c *StatusController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Info("Starting NetworkPolicyStatus controller")
+	defer klog.Info("Shutting down NetworkPolicyStatus controller")
+
+	klog.Info("Waiting for caches to sync for NetworkPolicyStatus controller")
+	if !cache.WaitForCacheSync(stopCh, c.cnpListerSynced, c.anpListerSynced) {
+		klog.Error("Unable to sync CNP caches for NetworkPolicy controller")
+		return
+	}
+	klog.Info("Caches are synced for NetworkPolicyStatus controller")
+
+	go wait.NonSlidingUntil(c.watchInternelNetworkPolicy, 5*time.Second, stopCh)
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *StatusController) watchInternelNetworkPolicy() {
+	watcher, err := c.internalNetworkPolicyStore.Watch(context.TODO(), "", labels.Everything(), fields.Everything())
+	if err != nil {
+		klog.Errorf("Failed to start watch for internal NetworkPolicy: %v", err)
+		return
+	}
+	defer watcher.Stop()
+	resultCh := watcher.ResultChan()
+	for {
+		select {
+		case event, ok := <-resultCh:
+			if !ok {
+				return
+			}
+			// Skip handling Bookmark events.
+			if event.Type == watch.Bookmark {
+				continue
+			}
+			np := event.Object.(*controlplane.NetworkPolicy)
+			if np.SourceRef.Type == controlplane.K8sNetworkPolicy {
+				continue
+			}
+			c.queue.Add(np.Name)
```

Comment:
Is it true we only care about span changes for an update?

Reply:
No, we also care about generation change, in which case the observedGeneration and phase should be updated.

Reply:
Could you add a comment to indicate what changes we care about here?